### PR TITLE
[CI] main2main-0717

### PR DESCRIPTION
**Pull request by zhao-stack**
#### Baselines

- Rebase the PR onto the latest vllm-ascend `main` commit [`69d1d6fe`](https://github.com/vllm-project/vllm-ascend/commit/69d1d6fe0cb15b07854bf058098a49d4165cc558).
- Keep the verified vLLM main target at [`fe784ff2`](https://github.com/vllm-project/vllm/commit/fe784ff22e630a31fd798f392b01e0a75c18f047), covering [`85c09e98...fe784ff2`](https://github.com/vllm-project/vllm/compare/85c09e9885e346ea1612da30ebff5a75f67d2350...fe784ff22e630a31fd798f392b01e0a75c18f047).
- Move the supported release lane from vLLM `v0.24.0` to `v0.25.0` ([`702f4814`](https://github.com/vllm-project/vllm/commit/702f4814fe54fabff350d43cb753ae3e47c0c276)).
- Final PR head: [`468fba5a`](https://github.com/vllm-project/vllm-ascend/commit/468fba5a975b0714f076f386ecdb93a253314168).

#### v0.25 compatibility decisions

| Area | Decision |
| --- | --- |
| Grouped block hashes (vLLM #45939) | v0.25 and the verified main target both use the terminal chained fine-grained hash. The old release-only rehash branch was removed instead of moving its gate to v0.25. |
| FLA relocation (vLLM #48500) | Keep a v0.25 gate: release imports `vllm.model_executor.layers.fla`, while the verified main target imports `vllm.third_party.flash_linear_attention`. |
| KV-cache contracts (vLLM #46384/#47782) | Keep v0.25 gates for manager-side hash resolution, exact hit lengths, allocation signatures, and coordinator return shapes. The per-group hit API returns a length tuple on both lanes, so that part is ungated; Recompute/Balance schedulers consume it with `max()`. Main-only partial Mamba-align hits use hash-block alignment. |
| GDN ABI (vLLM #46998) | Reuse the unified adapter already present on latest vllm-ascend main. Tests exercise the v0.25 two-argument/output-buffer call and the verified-main one-argument/tensor-return call; no duplicate production ABI branch is kept. |
| Qwen sequence parallel / logits (vLLM #47006/#48429/#48390) | Keep v0.25 gates because the release tag still has the older decoder, gather, and logits-head contracts. |
| MRV2 / speculative decode (vLLM #48261/#48167) | Target the verified main node only: use the unified graph manager, target buffers/groups, parameterless capture, and `_group_causal`. Model Runner V2 speculative decoding does not claim v0.25 tag compatibility until formal bring-up and debugging. |
| Hunyuan processor | Preserve the bundled v0.25 processor and clean its stale registry; use the native Transformers processor on verified main. Bare image-token wrapping is backported only to v0.25. |
| `swap_space` cleanup (vLLM #48549) | The removed test argument is compatible with both supported lanes, so no version gate is needed. |

#### Rebase notes

- The original PR branch was backed up before rewriting.
- The first rebase dropped the obsolete GDN production split because latest vllm-ascend main already contains the unified v0.25-compatible adapter from #12233.
- vllm-ascend main advanced once more during this update. The final rebase onto `69d1d6fe` was conflict-free, and `git range-diff` reported all 32 commits patch-equivalent to the already reviewed intermediate head.

### Does this PR introduce any user-facing change?

Yes. The supported release lane is now vLLM `v0.25.0`; the verified main lane remains pinned to `fe784ff2`. Model Runner V2 speculative decoding in this PR is intentionally main-only and does not claim release-tag compatibility.

### How was this patch tested?

- Ruff lint and format checks passed for the six files changed by the final MRV2 main-only cleanup.
- Focused mypy with the CI Python 3.10 settings passed for those six files; the previous `no-redef`, `valid-type`, and `misc` failures are eliminated.
- GitHub Actions `lint-and-select-tests` passed on head `5b2bccf7`, including both pre-commit and mypy.
- Follow-up head `468fba5a` keeps MRV2 speculative decoding main-only at the registration boundary: v0.25 skips only the Eagle, DFlash, and DSpark MRV2 speculative cases, while ordinary MRV2 coverage remains enabled. Focused Ruff, codespell, typos, repository checks, and mypy for Python 3.10/3.11/3.12 passed locally.
- `git diff --check` passed.
- The main2main mapping `validate` and `predict` passes were rerun against the exact verified target, followed by manual review of method, field, call, registration, inheritance, and broad candidates touched by this PR.
- Source contracts were compared directly at vLLM `v0.25.0` and `fe784ff2`. Non-MRV2 version gates above follow those exact refs; `worker/v2/spec_decode` is bound only to `fe784ff2`.
- Added/updated regression coverage for grouped hashes, KV-cache per-group hit tuples, GDN lane calls, FLA imports, and speculative graph registration.
- Local pytest collection is unavailable in this workspace because `torch` and `vllm` are not installed; head `468fba5a` has triggered GitHub CI for executable unit/device coverage.

- vLLM version: v0.25.0
- vLLM main: https://github.com/vllm-project/vllm/commit/85c09e9885e346ea1612da30ebff5a75f67d2350

### CODE REVIEW

**Comment by zhao-stack** [comment]
all E2E passed:
https://github.com/vllm-project/vllm-ascend/actions/runs/29566740482/job/87841686104?pr=12218

**Review by Tflowers-0129 — APPROVED** [review]

File: vllm_ascend/patch/worker/patch_v2/patch_eagle_speculator.py [hunk 1]
```diff
@@ -18,7 +18,6 @@
 
 from vllm.v1.worker.gpu.spec_decode.autoregressive import speculator as vllm_speculator_module
 
-from vllm_ascend.worker.v2.spec_decode.eagle.aclgraph import DecodeEagleAclGraphManager, PrefillEagleAclGraphManager
+from vllm_ascend.worker.v2.spec_decode.eagle.aclgraph import EagleAclGraphManager
```
Comment (drslark) [inline vllm_ascend/patch/worker/patch_v2/patch_eagle_speculator.py:21]: Did vllm deletes `PrefillEagleAclGraphManager`?
Now, there is only `DecodeEagleAclGraphManager`?

File: vllm_ascend/worker/v2/spec_decode/eagle/aclgraph.py [hunk 2]
```diff
@@ -164,35 +73,45 @@ def capture(
...
+                    skip_attn=(desc.cg_mode == CUDAGraphMode.PIECEWISE),
+                )
+                return lambda cg_mode: forward_fn(
+                    num_reqs,
+                    cg_mode == CUDAGraphMode.PIECEWISE,
+                    BatchExecutionDescriptor(cg_mode=cg_mode, num_tokens=num_tokens, num_reqs=num_reqs),
+                    num_tokens_across_dp,
+                )
+
+            CudaGraphManager.capture(self, create_forward_fn, progress_bar_desc=progress_bar_desc)
```
Comment (drslark) [inline vllm_ascend/worker/v2/spec_decode/eagle/aclgraph.py:120]: Did vllm change `capture` to `staticmethod`?

File: vllm_ascend/patch/worker/patch_v2/patch_eagle_speculator.py (hunk 1 above)
Comment (zhao-stack) [reply vllm_ascend/patch/worker/patch_v2/patch_eagle_speculator.py:21]: DecodeEagleAclGraphManager and PrefillEagleAclGraphManager are merged into EagleAclGraphManager.

File: vllm_ascend/worker/v2/spec_decode/eagle/aclgraph.py (hunk 2 above)
Comment (zhao-stack) [reply vllm_ascend/worker/v2/spec_decode/eagle/aclgraph.py:120]: No, CudaGraphManager.capture is still an instance method. The explicit call
CudaGraphManager.capture(self, create_forward_fn, ...)
is used to bypass SpeculatorCudaGraphManager.capture for the Ascend decode path. This path needs its own create_forward_fn to preserve the Ascend-specific PIECEWISE skip_attn handling and num_tokens_across_dp preparation.
self is therefore passed explicitly as with a normal unbound instance-method call. For the prefill path, we use super().capture(...) directly to reuse the new unified upstream implementation.

File: vllm_ascend/worker/v2/spec_decode/eagle/aclgraph.py [hunk 3]
```diff
@@ -29,99 +30,7 @@
...
-                self.vllm_config,
-                self.speculator.speculative_config,
-                positions.shape[0],
-                draft_attn_metadatas=draft_attn_metadatas,
-            )
-        return ret
-
-
-class DecodeEagleAclGraphManager(DecodeSpeculatorCudaGraphManager):
+class EagleAclGraphManager(SpeculatorCudaGraphManager):
```
Comment (wxsIcey) [inline vllm_ascend/worker/v2/spec_decode/eagle/aclgraph.py:33]: I reviewed the logic, merging the two Managers here is right. However, I think naming the class `SpeculatorAclGraphManager` would be more appropriate. This class will be refactored and promoted to the autoregressive directory later.

File: vllm_ascend/patch/worker/patch_v2/patch_eagle_speculator.py (hunk 1 above)
Comment (wxsIcey) [reply vllm_ascend/patch/worker/patch_v2/patch_eagle_speculator.py:21]: Yes, vllm delete it, please see https://github.com/vllm-project/vllm/pull/48261 @drslark

**Review by wxsIcey — APPROVED** [review]
I think the modification to the mrv2 spec decode is correct. For other modules, please confirm with the relevant person in charge.

File: vllm_ascend/patch/worker/patch_v2/patch_eagle_speculator.py (hunk 1 above)
Comment (drslark) [reply vllm_ascend/patch/worker/patch_v2/patch_eagle_speculator.py:21]: Okay, but I think the name `SpeculatorCudaGraphManager` should be changed to `EagleAclGraphManager`.

**Review by drslark — CHANGES_REQUESTED** [review]

File: vllm_ascend/worker/v2/spec_decode/eagle/aclgraph.py (hunk 2 above)
Comment (drslark) [reply vllm_ascend/worker/v2/spec_decode/eagle/aclgraph.py:120]: Okay, i see.
You are right.
I think it is okay.

**Review by jiajinzhu2 — APPROVED** [review]
I reviewed the speculative decoding section. LGTM!

File: vllm_ascend/worker/v2/spec_decode/eagle/speculator.py
Comment (zzzzwwjj) [inline vllm_ascend/worker/v2/spec_decode/eagle/speculator.py:186]: Why is there still has an object named `prefill_cudagraph_manager`?

**Review by zzzzwwjj — APPROVED** [review]

**Comment by DreamerLeader** [comment]
I have already reviewed the code pooling section.LGTM!

File: vllm_ascend/worker/v2/spec_decode/eagle/speculator.py
Comment (zhao-stack) [reply vllm_ascend/worker/v2/spec_decode/eagle/speculator.py:186]: The manager classes were merged, but the prefill and decode manager instances were not merged into a single object. Upstream still creates two SpeculatorCudaGraphManager instances and stores them in self.prefill_cudagraph_manager and self.decode_cudagraph_manager:
https://github.com/vllm-project/vllm/blob/fe784ff22e630a31fd798f392b01e0a75c18f047/vllm/v1/worker/gpu/spec_decode/autoregressive/speculator.py#L48-L84
They use different configurations and serve different paths: the prefill manager uses decode_query_len = num_speculative_steps + 1, while the decode manager uses decode_query_len = 1 and a different CUDA graph mode. They also capture different functions and input buffers.
In vLLM Ascend, the patched SpeculatorCudaGraphManager resolves to EagleAclGraphManager, so super().init_cudagraph_manager() creates two separate EagleAclGraphManager instances while retaining the upstream attribute names. Therefore, self.prefill_cudagraph_manager is still expected here.

**Review by DreamerLeader — APPROVED** [review]
I reviewed the speculative decoding section. LGTM!

**Review by MengqingCao — APPROVED** [review]
plz drop the support of 0.24.0 in the follow-up pr.

**Comment by wangxiyuan** [comment]
yes, we should upgrade to 0.25.0 first https://github.com/vllm-project/vllm-ascend/pull/12233

**Review by linfeng-yuan — APPROVED** [review]

File: vllm_ascend/patch/worker/patch_v2/patch_eagle_speculator.py (hunk 1 above)
Comment (drslark) [reply vllm_ascend/patch/worker/patch_v2/patch_eagle_speculator.py:21]: My mistake, it is `EagleAclGraphManager` to be change to `SpeculatorCudaAclManager`.
Also, this task should not be done in this pr.
I will add it to my todo list.

Comment (drslark) [reply vllm_ascend/patch/worker/patch_v2/patch_eagle_speculator.py:21]: LGTM. I think it is good. I will approve it.
Thanks for your adaptation.

**Review by drslark — APPROVED** [review]

**Comment by zhao-stack** [comment]
This PR will not be merged. We will use the main2main workflow to upgrade.
